kconfig: Add symbols for C standard

pdgendt opened the pull request:
Add symbols to select a C standard. Next to an option, hidden symbols are available to select the minimum.
This allows subsystems or modules to select the minimum compliant C standard supported.

pdgendt committed:
kconfig: Add symbols for C standard

Add symbols to select a C standard. Next to an option, hidden symbols
are available to select the minimum.
This allows subsystems or modules to select the least compliant C
standard supported.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
arch: posix: Select at least C11 standard

Replace the global CSTD property with the CSTD kconfig option to select
at least C11 standard.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

aescolar reviewed (commented):
Overall this seems fine to me. My only doubt is about having a new `lib/c` folder just for this (which would be orphaned maintainer wise)

pdgendt commented:
> Overall this seems fine to me. My only doubt is about having a new `lib/c` folder just for this (which would be orphaned maintainer wise)

I was looking for a logical place to put it, but didn't find it obvious. If a more suitable place can be suggested, I can move it.

aescolar commented:
> > Overall this seems fine to me. My only doubt is about having a new `lib/c` folder just for this (which would be orphaned maintainer wise)
> 
> I was looking for a logical place to put it, but didn't find it obvious. If a more suitable place can be suggested, I can move it.

What about Kconfig.zephyr ? Maybe inside the compiler options menu?

tejlmand commented on Kconfig.zephyr:
```diff
@@ -338,6 +338,71 @@ endmenu
 
 menu "Compiler Options"
 
+config STD_AT_LEAST_C99
+	bool
+	help
+	  Hidden option to select compiler support C99 standard or higher.
```
Would a better wording / name be `MINIMUM_REQUIRED_STD_C99` ?

Initially when reading `AT_LEAST` I got to think of something which the compiler / toolchain could be setting.

tejlmand commented on CMakeLists.txt:
```diff
@@ -947,8 +947,23 @@ if(CONFIG_USERSPACE)
   set(PROCESS_GPERF ${ZEPHYR_BASE}/scripts/build/process_gperf.py)
 endif()
 
+# Backwards compatible: allow manual selection of C standard using a global property
```
I think we could deprecated the use of `CSTD` and print a deprecation warning when defined.

tejlmand reviewed (commented):
Looks good, thanks for this proposal.

Two minor improvement proposals.

pdgendt committed:
arcmwdt: Enable GNU C Extensions

Replace the global CSTD property with a Kconfig symbol selection.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

aescolar commented:
> I think we could deprecated the use of CSTD and print a deprecation warning when defined.

If we do this, I think we need a sentence in the migration guide and release notes.

pdgendt committed:
doc: releases: release-notes: 3.7: mention deprecation of CSTD

Add an entry that the CSTD cmake property is deprecated and replaced
with a kconfig choice.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
Updated as per comments, and removed the `zcbor` module change for now, as this can go in a separate PR.

57300 commented on CMakeLists.txt:
```diff
@@ -948,7 +948,24 @@ if(CONFIG_USERSPACE)
+  elseif(CONFIG_STD_C90)
+    set(CSTD c90)
+  else()
+    assert(0 "Unreachable code. Expected C standard to have been chosen.")
```
nit: `assert` is a Zephyr-specific macro, which isn't needed here.
```suggestion
    message(FATAL_ERROR "Unreachable code. Expected C standard to have been chosen.")
```

57300 commented on CMakeLists.txt:
```diff
@@ -948,7 +948,24 @@ if(CONFIG_USERSPACE)
+  endif()
+else()
+  message(DEPRECATION
+    "Global CSTD property is deprecated, see Kconfig.zephyr for C Standard options.")
```
Technically, this is deprecating not just the global property, but also passing `-DCSTD=...` as seen here:
https://github.com/zephyrproject-rtos/zephyr/blob/fd1a129a3e6330bdf44324e47ac7638dd6cdf5a5/tests/subsys/bindesc/definition/testcase.yaml#L24-L35

57300 replied:
> should we introduce a `CONFIG_STD_C_EXTENSIONS` too for the gnu extensions?

Sure, if @tejlmand agrees.

I'm not too sure how C standard selection should work, because I'm only learning now that CMake lets you control this on a per-target basis, using [C_STANDARD](https://cmake.org/cmake/help/latest/prop_tgt/C_STANDARD.html) and [C_EXTENSIONS](https://cmake.org/cmake/help/latest/prop_tgt/C_EXTENSIONS.html) properties. Maybe a global setting is not needed?

pdgendt replied:
I have to admit, this is kind of getting over my head 🙃
I tried with `C_STANDARD` and `C_EXTENSIONS` but this doesn't make the `-std=` compiler argument explicit it seems, and fails in some tests.

pdgendt committed:
cmake: Replace assert with fatal error message

Use vanilla cmake FATAL_ERROR messages instead of assert.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt commented:
Hey @tejlmand @57300, thanks for your comments already.

I've reverted to a simpler version, some notes:

- Using `CMAKE_C_STANDARD` seems complicated as this needs to be set before any target is created
- Maybe we shouldn't deprecate `CSTD` if the user would like to set a specific version with extensions?
- If CSTD would be deprecated, how should we allow `C_EXTENSIONS` to be enabled? Currently the compiler flag is set, as is
- I kind of prefer to set the standard globally instead of on target basis, but I could be alone in that

stephanosio reviewed (changes requested):
Just giving it a -1 until we reach a consensus on how build system/toolchain configurations, such as C and C++ standard selection, should be done.

@tejlmand I vaguely remember discussing this in the Toolchain WG a few years back -- based on that, it was my understanding that your preference was to keep the build system/toolchain configurations within CMake as CMake variables so that they are available from the very beginning of the build cycle (i.e. before Kconfig runs), and that `CONFIG_CPP_STDxx` Kconfigs are the oddballs that need to be fixed. Has this view changed?

> I'm not too sure how C standard selection should work, because I'm only learning now that CMake lets you control this on a per-target basis, using [C_STANDARD](https://cmake.org/cmake/help/latest/prop_tgt/C_STANDARD.html) and [C_EXTENSIONS](https://cmake.org/cmake/help/latest/prop_tgt/C_EXTENSIONS.html) properties.

@57300 I believe using `C_STANDARD` (or `CXX_STANDARD` for that matter) is not a viable option because it requires CMake to be fully aware of the flags and features supported by a compiler, which may not always be the case -- that is why we have our own [build system "compiler abstraction layer"](https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/compiler/compiler_flags_template.cmake) in the first place.

> Maybe a global setting is not needed?

I believe a global setting is still useful because it can be used to set the "default" setting -- per-target setting can be set to override the global/default setting.

IMO, it is better to keep the C/C++ standard settings within the build system because, that way, 1) these settings are available from the very beginning of the build process and can be consumed by anything that runs before Kconfig, 2) per-target configuration can be supported (though, this requires additional work at this time).

If we are redoing this, my preference would be to:

* Declare `STD_CPPxx` Kconfigs a mistake and deprecate them.
* Implement a new mechanism to streamline global and per-target C/C++ standard selection (e.g. `zephyr_c_std(c99)` for global, `zephyr_library_c_std(c99)` for per-library/target).

@tejlmand What do you think?

dkalowsk commented:
> IMO, it is better to keep the C/C++ standard settings within the build system because, that way, 1) these settings are available from the very beginning of the build process and can be consumed by anything that runs before Kconfig, 2) per-target configuration can be supported (though, this requires additional work at this time).

I would suggest that the Zephyr code have a minimum supported version for both C and C++.  As time moves on, it's likely that there will be some backwards compatibility breaking change that happens in either one of the standards languages and some code will make use of it.  Having the Zephyr source declaring it's minimum support will help prevent lots of headaches when/if this happens.

> If we are redoing this, my preference would be to:
> 
> * Declare `STD_CPPxx` Kconfigs a mistake and deprecate them.

Keeping C and C++ configurations the same would be a good thing, and deprecating these Kconfigs makes the most sense.  It's not providing a lot of benefit to have to regularly update them with C++2c or whatever the new standard is.

> * Implement a new mechanism to streamline global and per-target C/C++ standard selection (e.g. `zephyr_c_std(c99)` for global, `zephyr_library_c_std(c99)` for per-library/target).

tejlmand commented:
> @tejlmand I vaguely remember discussing this in the Toolchain WG a few years back -- based on that, it was my understanding that your preference was to keep the build system/toolchain configurations within CMake as CMake variables so that they are available from the very beginning of the build cycle (i.e. before Kconfig runs), and that CONFIG_CPP_STDxx Kconfigs are the oddballs that need to be fixed. Has this view changed?

@stephanosio we had discussions on toolchain features / testing of features instead of relying on Kconfig to describe a given toolchain feature set.

Such testing can then provide input to Kconfig, from where different settings can be enable.
This is partly because current hardcoded `TOOLCHAIN_HAS_<feature>` is not really scalable, especially not with 3rd party toolchains. As example we can handle newlib and picolibc support in Zephyr SDK, but not in other 3rd party toolchains.

The selection of a given feature (if toolchain reports it's supported) can then still be done in Kconfig.
See MoM 03 October 2022: https://docs.google.com/document/d/11ENuzr2oDsHXa2FrHf0S3AQ2HncQbGBCU9F5sjTZUnI

So view has not changed 🙂 
And with the choice and a future toolchain feature test scheme, then choice entry could become:
```
config STD_C99
	bool "C99"
	depends on !MINIMUM_REQUIRED_STD_C11 && TOOLCHAIN_SUPPORT_C99
```

> 1) these settings are available from the very beginning of the build process and can be consumed by anything that runs before Kconfig, 2) per-target configuration can be supported (though, this requires additional work at this time)

It is today the toolchain (compiler) which defines how a given feature maps to a compiler flag using the compiler property target.
But it is the Kconfig which defines if a given feature should be enabled:
As example, look here: https://github.com/zephyrproject-rtos/zephyr/blob/main/CMakeLists.txt#L289-L313

This works well, because a toolchain can clear the property if it is not supported.
The downside is that it's not reflected in Kconfig, meaning a user might think featureX can be enabled, but toolchain doesn't support it and thus no argument is passed to the compiler.

tejlmand commented:
From Toolchain WG meeting 2024-04-15:
- Keeping C standard selection in Kconfig as choice.
   This allows a sample to select a given standard if there is a need, while also ensure that the minimum required selecting in a subsystem will cause a Kconfig error if there is a mismatch between minimum required and the sample config
- Extend with a simple config in Kconfig to allow users / samples to request the use of GNU extensions
   Toolchains shall be able to identify if they don't support GNU extensions, so it must be guarded with a `depends on TOOLCHAIN_SUPPORTS_GNU_EXTENSIONS`
- Later, when toolchain feature testing is implemented, the `TOOLCHAIN_SUPPORTS_x` should be provided as result of such testing.

pdgendt committed:
kconfig: Add symbols for GNU C Extensions

Add a symbol to enable GNU C Extensions. And a hidden option for
toolchains to signal GNU Extensions support.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

pdgendt committed:
tests: bindesc: Replace CSTD definition with Kconfig symbols

Replace the bindesc testcase CSTD definitions with the newly introduced
kconfig symbols.

Signed-off-by: Pieter De Gendt <pieter.degendt@basalte.be>

tejlmand commented on Kconfig.zephyr:
```diff
@@ -403,6 +403,19 @@ config STD_C23
 
+config TOOLCHAIN_SUPPORTS_GNU_EXTENSIONS
+	bool
+	default y if "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr" || "$(ZEPHYR_TOOLCHAIN_SUPPORTS_GNU_EXTENSIONS)" = "y"
```
what is: `ZEPHYR_TOOLCHAIN_SUPPORTS_GNU_EXTENSIONS` ?

Also we should have an update here: https://github.com/zephyrproject-rtos/sdk-ng/blob/main/cmake/zephyr/Kconfig:
```
config TOOLCHAIN_ZEPHYR_SUPPORTS_GNU_EXTENSIONS
	def_bool y
	select TOOLCHAIN_SUPPORTS_GNU_EXTENSIONS
```
but the `default y if "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "zephyr"` can stay as older Zephyr Toolchains also support gnu extensions.

pdgendt replied:
I did the same as was done here:
https://github.com/zephyrproject-rtos/zephyr/blob/5919bd988b9dfb40a6b4ab5a3fd4af2dccda3c95/kernel/Kconfig#L1009-L1014

pdgendt replied:
I've created a PR: https://github.com/zephyrproject-rtos/sdk-ng/pull/746

tejlmand replied:
> I did the same as was done here:

which seems to have been introduced at a point when I was away 😄 
https://github.com/zephyrproject-rtos/zephyr/pull/45656

You already added the Kconfig to indicate when cross-compile supports gnu extension or not.
And we should be careful not to introduce too many ways to achieve the same (leads to confusion), but in this particular case, I acknowledge that especially on Linux similar, setting this as env is convenient,  but the caveat is that users then forget to clear if switching to a toolchain not supporting such extensions.

So when picking up from environment, then we should only adjust default for the cross-compile, as to not impact all toolchains. Seems I have some followup to do on the `TLS` part.

But I see the setting has been removed, which I think is good for time being.

tejlmand reviewed (commented):
looks good.

But fail to understand the reason for the `ZEPHYR_TOOLCHAIN_SUPPORTS_GNU_EXTENSIONS` setting.

stephanosio commented on Kconfig.zephyr:
```diff
@@ -338,6 +338,84 @@ endmenu
 
 menu "Compiler Options"
 
+config MINIMUM_REQUIRED_STD_C99
+	bool
+	help
+	  Hidden option to select compiler support C99 standard or higher.
```
would be better to name these `REQUIRES_STD_Cxx` to align with the existing symbols of the same category (e.g. `REQUIRES_FULL_LIBC`, `REQUIRES_FULL_LIBCPP`, `REQUIRES_FLOAT_PRINTF`).

tejlmand replied:
nicely spotted.

stephanosio reviewed (commented):
No objection to this as long as the plan is to use Kconfig for handling toolchain capabilities and configurations.

Just a minor nit about the Kconfig naming.
